tools/ci/java-ci-tools/src/main/java/org/apache/flink/tools/ci/licensecheck/LicenseChecker.java

```diff
@@ -0,0 +1,320 @@
+		}
+
+		// collect all declared dependencies from NOTICE file
+		Set<IncludedDependency> declaredDependencies = new HashSet<>();
+		try (BufferedReader br = new BufferedReader(new StringReader(noticeContents))) {
+			String line;
+			while ((line = br.readLine()) != null) {
+				Matcher noticeDependencyMatcher = NOTICE_DEPENDENCY_PATTERN.matcher(line);
+				if (noticeDependencyMatcher.find()) {
+					String groupId = noticeDependencyMatcher.group(1);
+					String artifactId = noticeDependencyMatcher.group(2);
+					String version = noticeDependencyMatcher.group(3);
+					if (groupId == null && artifactId == null && version == null) { // "bundles" case
+						groupId = noticeDependencyMatcher.group(5);
+						artifactId = noticeDependencyMatcher.group(6);
+						version = noticeDependencyMatcher.group(7);
+					}
+					IncludedDependency toAdd = IncludedDependency.create(groupId, artifactId, version);
+					if (!declaredDependencies.add(toAdd)) {
+						LOG.warn("Dependency {} has been declared twice in module {}", toAdd, moduleName);
+					}
+				}
+			}
+		}
+		// print all dependencies missing from NOTICE file
+		Set<IncludedDependency> expectedDependencies = new HashSet<>(modulesWithShadedDependencies.get(moduleName));
+		expectedDependencies.removeAll(declaredDependencies);
+		for (IncludedDependency missingDependency : expectedDependencies) {
+			LOG.error("Could not find dependency {} in NOTICE file {}", missingDependency, noticeFile);
+			severeIssueCount++;
+		}
+
+		if (!MODULES_DEFINING_EXCESS_DEPENDENCIES.contains(moduleName)) {
+			// print all dependencies defined in NOTICE file, which were not expected
+			Set<IncludedDependency> excessDependencies = new HashSet<>(declaredDependencies);
+			excessDependencies.removeAll(modulesWithShadedDependencies.get(moduleName));
+			for (IncludedDependency excessDependency : excessDependencies) {
+				LOG.warn("Dependency {} is mentioned in NOTICE file {}, but is not expected there", excessDependency, noticeFile);
+			}
+		}
+
+		return severeIssueCount;
+	}
+
+	private static String readFile(Path path) throws IOException {
+		byte[] encoded = Files.readAllBytes(path);
+		return new String(encoded, Charset.defaultCharset());
+	}
+
+	private List<File> findNoticeFiles(File root) throws IOException {
+		return Files.walk(root.toPath())
+			.filter(file -> {
+				int nameCount = file.getNameCount();
+				return file.getName(nameCount - 1).toString().equals("NOTICE")
+					&& file.getName(nameCount - 2).toString().equals("META-INF")
+					&& file.getName(nameCount - 3).toString().equals("resources");
+			})
+			.map(Path::toFile)
+			.collect(Collectors.toList());
+	}
+
+	private Multimap<String, IncludedDependency> parseModulesFromBuildResult(File buildResult) throws IOException {
+		Multimap<String, IncludedDependency> result = ArrayListMultimap.create();
+		try (BufferedReader br = new BufferedReader(new FileReader(buildResult))) {
```

Comment:
If you were to use `Files.readAllLines()` you could use an indexed for loop, and parse each block within a single iteration, instead of having to keep state around to modify the behavior of the next one.

Reply:
I agree that this is an alternative approach. It would imply materializing the entire build output as strings into the JVM. I'm not sure if the code is a lot more readable when having multiple nested loops fiddling with the line indexes.
For the sake of my time, I would like to keep the code as-is for now. I agree it is not perfectly easy to understand, but the improvement wouldn't be substantial enough to pay the performance and reimplementation price.

Reply:
Fair enough. (Note that you could use Files.lines() instead of a BufferedReader, which should make it overall a tiny bit simpler)